### PR TITLE
Document test exception handling

### PR DESCRIPTION
**Pull request by driptaroop**
## Overview

Documents how exceptions in tests are handled. This PR contains documentation that describes,

1. Uncaught exceptions cause test failures by default.
2. How assertions are implemented using exceptions.
3. JUnit 5 does not differentiate between failed assertions and other exceptions but IDEs and other tools may choose to do so by checking for instances of AssertionError.
4. Adds assertDoesNotThrow example for Java.

Solves Issue: #3730
---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### CODE REVIEW

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc
Comment (sbrannen) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:397]: Add note that other assertion libraries provide similar support, such as AssertJ's `assertThatNoException().isThrownBy(() -> ...)`.

File: documentation/src/test/java/example/exception/AssertDoesNotThrowExceptionDemoTestCase.java
Comment (sbrannen) [inline documentation/src/test/java/example/exception/AssertDoesNotThrowExceptionDemoTestCase.java:23]: We use static imports for `Assertions`.

File: documentation/src/test/java/example/exception/AssertDoesNotThrowExceptionDemoTestCase.java
Comment (sbrannen) [inline documentation/src/test/java/example/exception/AssertDoesNotThrowExceptionDemoTestCase.java:27]: Please move this to a separate example.

**Review by sbrannen — CHANGES_REQUESTED** [review]
Thanks for the PR.

I've requested a few changes.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc
Comment (Marcono1234) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:403]: Question to the maintainers: Would it make sense to also mention `assertThrows` in a small section?

I know that `assertThrows` is used in the existing code samples, for example in the "Assertions" section. But users might first have a look at this new section and are then confused because there is no information about handling expected exceptions, only about unexpected exceptions. This might be especially a problem when this section is named "Exception Handling" (as proposed in https://github.com/junit-team/junit5/pull/3780#discussion_r1567605519).

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc
Comment (Marcono1234) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:385]: Question to the maintainers: Should this also explicitly mention that a `throws` clause has no effect, to avoid any misunderstanding?
For example
> [...] It does not matter whether or not the test method has a `throws` clause which lists the exception (or a superclass of it).

Otherwise users might assume `throws SomeException` means that it is expected that the test fails with `SomeException`.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc
Comment (sbrannen) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:403]: Yes, I think we should cover `assertThrows` (and mention `assertThrowsExactly`) in this new section.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc
Comment (sbrannen) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:385]: Standard Java semantics apply here, but that's not always readily apparent to someone new to testing.

So, yes, it would be beneficial for some users to point that out.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc
Comment (driptaroop) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: I agree. changed it.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc
Comment (driptaroop) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: Added the "see also" section with "Third-party Assertion Libraries"

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc
Comment (driptaroop) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: sorry I missed this one first because I had soft wrapping enabled. did it now.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc
Comment (driptaroop) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: made the anchors more hierarchical

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc
Comment (driptaroop) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: added assertJ's non-exception assertion in this section

File: documentation/src/test/java/example/exception/AssertDoesNotThrowExceptionDemoTestCase.java
Comment (driptaroop) [reply documentation/src/test/java/example/exception/AssertDoesNotThrowExceptionDemoTestCase.java]: done

File: documentation/src/test/java/example/exception/AssertDoesNotThrowExceptionDemoTestCase.java
Comment (driptaroop) [reply documentation/src/test/java/example/exception/AssertDoesNotThrowExceptionDemoTestCase.java]: I created separate sections for `assertThrows`, `assertThrowsExactly`, and `assertDoesNotThrow` and moved individual examples accordingly.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc
Comment (driptaroop) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:460]: created separate sections for `assertThrows`, `assertThrowsExactly`, and `assertDoesNotThrow`

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc
Comment (driptaroop) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: added a section describing the throws clause in junit exception context

File: documentation/src/test/java/example/exception/AssertDoesNotThrowExceptionDemoTestCase.java
Comment (sbrannen) [inline documentation/src/test/java/example/exception/AssertDoesNotThrowExceptionDemoTestCase.java:20]: I realize this is not officially documented, but test classes in the `documentation` module must be named `*Demo`.

https://github.com/junit-team/junit5/blob/3e03c97f83d1ff35dcb2b3a7cc48b54943eb1b69/documentation/documentation.gradle.kts#L160

So, please rename all `*TestCase` example classes to `*Demo` so that they are run with the Gradle build.

**Review by sbrannen — CHANGES_REQUESTED** [review]
Thanks for making all of the requested changes.

This is looking very good. 👍 

I just added one (hopefully final) request to rename the example test classes.

File: documentation/src/test/java/example/exception/AssertDoesNotThrowExceptionDemoTestCase.java
Comment (driptaroop) [reply documentation/src/test/java/example/exception/AssertDoesNotThrowExceptionDemoTestCase.java:20]: Thanks for the suggestion. I initially named the classes as `*Demo` but that was failing the build because of `UncaughtExceptionHandlingDemoTestCase` (which was designed to fail). So I renamed it and others (to keep parity) to `*TestCase` to exclude from build. If I rename then to `*Demo` then the build will fail. 

One thing I can do is I can rename all the tests except `UncaughtExceptionHandlingDemoTestCase` to make the builds pass. do you think that's a good idea? do you have any other suggestions?

Comment (sbrannen) [reply documentation/src/test/java/example/exception/AssertDoesNotThrowExceptionDemoTestCase.java:20]: > do you have any other suggestions?

Yes. 😉 

Please name them all `*Demo` and make use of our `@ExpectToFail` annotation/extension.

See `example.AssertionsDemo.timeoutExceeded()` for details.

Comment (sbrannen) [reply documentation/src/test/java/example/exception/AssertDoesNotThrowExceptionDemoTestCase.java:20]: https://github.com/junit-team/junit5/blob/3e03c97f83d1ff35dcb2b3a7cc48b54943eb1b69/documentation/src/test/java/example/AssertionsDemo.java#L131-L143

Comment (driptaroop) [reply documentation/src/test/java/example/exception/AssertDoesNotThrowExceptionDemoTestCase.java]: ah this is nice and exactly what I wanted. Thanks for the suggestion. I have changed all the classes name to `*Demo`

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc
Comment (Marcono1234) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:421]: Maybe it would be useful to mention that `assertThrows` and `assertThrowsExactly` return the thrown exception to allow performing additional assertions on it, such as checking its message?
And then maybe include an example for this in `ExceptionAssertionDemo` and `ExceptionAssertionExactDemo` (or adjust the existing ones)?

File: documentation/src/test/java/example/exception/ExceptionAssertionExactDemo.java
Comment (Marcono1234) [inline documentation/src/test/java/example/exception/ExceptionAssertionExactDemo.java:29]: This message looks incorrect? The test will actually pass because exactly the same exception type is thrown.

File: documentation/src/test/java/example/exception/ExceptionAssertionExactDemo.java
Comment (Marcono1234) [inline documentation/src/test/java/example/exception/ExceptionAssertionExactDemo.java:33]: To the maintainers: Would it make sense to rather have this "This will fail because ..." as comment instead of as message? Because as message it seems a bit weird.

Comment (sbrannen) [reply documentation/src/test/java/example/exception/ExceptionAssertionExactDemo.java:33]: Good point, @Marcono1234.

Like with the timeout assertion examples in the [Assertions](https://junit.org/junit5/docs/current/user-guide/#writing-tests-assertions) section, let's move all "comments" to actual code comments preceding the corresponding assertions, and let's do that consistently across all examples introduced in this PR.

Consequently, let's remove the custom failure messages completely. If we later come up with succinct failure messages that make sense in the examples (which is likely difficult due to the simplicity of the examples), we can add them after this PR, but I'd like to get this PR finished and merged soon (potentially in time for 5.11 M1).

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc
Comment (sbrannen) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:421]: Yes, I think it's worth mentioning that explicitly.

The `exceptionTesting()` example in the Assertions section already demonstrates how to check the exception message in the returned exception, but I agree that it's worth repeating that in the `assertThrows` and `assertThrowsExactly` examples in these new sections.

File: documentation/src/test/java/example/exception/ExceptionAssertionDemo.java
Comment (Marcono1234) [inline documentation/src/test/java/example/exception/ExceptionAssertionDemo.java:32]: The arguments should be switched; these methods define left as _expected_ and right as _actual_, see the [`assertEquals` documentation](https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/Assertions.html#assertEquals(java.lang.Object,java.lang.Object)).

Otherwise the assertion error thrown on mismatch will be quite confusing.
```suggestion
        assertEquals("Expected exception", exception.getMessage());
```

File: documentation/src/test/java/example/exception/ExceptionAssertionExactDemo.java
Comment (Marcono1234) [inline documentation/src/test/java/example/exception/ExceptionAssertionExactDemo.java:35]:
```suggestion
        assertEquals("Expected exception", exception.getMessage());
```

**Review by sbrannen — APPROVED** [review]
Thanks for making all of the requested changes.

This looks good now.

I'll perform any further polishing after merging this PR.

**Comment by driptaroop** [comment]
thank you for the suggestions and changes. Afaik, you will merge this and there is nothing else for me to do right?

**Comment by sbrannen** [comment]
> Afaik, you will merge this and there is nothing else for me to do right?

Correct

**Comment by sbrannen** [comment]
This has been merged into `main` in f915d67f186693b9a98a0cbead6707af3da6d8ea and revised in 27f3b7cf68d268721ca32edd11cdd175a51afa35.

Thanks for the PR and the great collaboration! 👍

**Pull request closed without merging**
